doc/rpl.5
.TH "rpl" "5" "2009-09-28" "ttyrpld" "tty logging daemon suite"
.SH "Name"
.PP
rpl(5) \(em ttyrpld log file format
.SH "Description"
.PP
The logfiles rpld creates are a small variation of the packet protocol as
described in rpldev(4). The \fIdev\fP field is not present. (Since ttyrpld
v2.00, the timestamp is already added in the kernel.) All fields are little
endian and packed, i.e. there are no alignment gaps. The structure is similar
to the struct rpldev_packet:
.PP
.nf
struct rpldsk_packet {
	union rpldev_evmagic evmagic;
	uint32_t size;
	struct rpltime time;
} __attribute__((packed));
.fi
.PP
Possible values for \fIevent\fP equal those listed in rpldev(4), plus the
following:
.IP \(bu 4
RPLEVT_ID_PROG = 0x72703345, /* "rp3E" */
.IP \(bu 4
RPLEVT_ID_DEVPATH = 0x72703346, /* "rp3F" */
.IP \(bu 4
RPLEVT_ID_TIME = 0x72703347, /* "rp3G" */
.IP \(bu 4
RPLEVT_ID_USER = 0x72703348, /* "rp3H" */
.PP
A RPLEVT_ID_PROG is added to each logfile by rpld to contain the program with
which it was created. (rpld in this case \(em but anyone is free to write a
different daemon.)
.PP
A RPLEVT_ID_DEVPATH contains the path of the device node that has been tapped.
There is a special handling case in rpld when multiple device nodes with the
same major-minor number are used at the same time:
.IP \(bu 4
if the first packet rpld picks up for a certain ->dev line (tty) is a
RPLEVT_INIT or RPLEVT_OPEN packet with a dentry name (e.g. /dev/tty1), this
name will be used for this major-minor number
.IP \(bu 4
rpld will look into /dev and take the first pick
.PP
In the default case, you should not worry, as each major-minor number only has
one node in /dev, and even if it does not, you are able to know what tty was
logged, because the dentries are similarly-named.
.PP
RPLEVT_ID_TIME packets carry the time the log was started. This is useful
because the logfile's timestamp on the filesystem may change due to user
interaction.
.PP
Last, but not least, RPLEVT_ID_USER contains the username (or UID if no user
was found) that was being traced. Especially useful when users get removed
after the logs have been recorded.
.PP
All of these four packets are informational ones and are not required for
proper replay with ttyreplay.
.SH "See also"
.PP
\fBrpldev\fP(4)


doc/changelog.txt

Notes
=====

You do not need to repatch your kernel or recompile rpldev if they have not
changed significantly between ttyrpld releases.

│                │last time kpatch│                            │
│Operating system│(rpldhk) changed│last time rpldev changed:   │
├────────────────┼────────────────┼────────────────────────────┤
│Linux           │2.50            │2.60                        │
│FreeBSD         │2.60            │2.60                        │
│OpenBSD         │-               │2.60                        │
│NetBSD          │2.50            │2.60                        │


Changelog
=========

HEAD

Changes:
- doc: move changelog to text/plain
- rpldev: NetBSD 5.0.1 support
- rpldev: OpenBSD 4.5 support
- rpldev: Linux 2.6.31 support
- rpldev: FreeBSD 8.0 support
- packet: increase the "size" field to 32 bits
- packet: combine "magic" and "event" fields, and move to front


v2.52 (2008-09-07)

  • Updated kernel components for Linux 2.6.27
  • Updated userspace code for libHX 1.25

v2.51 (2008-05-15)

  • Updated rpldhk and rpldev for Linux 2.6.25, OpenBSD 4.3, FreeBSD 7.0,
    NetBSD 4.0

v2.50 (2008-01-01)

  • changed the packet format to be compatible between 32-bit and 64-bit
  • added rplcvt utility for converting old on-disk rpld files
  • explicitly blacklist /dev/stdout, /dev/stderr and /dev/stderr - these have
    been popping up as tty names because there is such a symlink in Linux's /
    dev. (But we need symlink processing for Solaris, eww.)

v2.19 (2007-09-17)

  • The kernel interface has been reduced (hooks removed). The 2.19 userspace
    is still compatible (throws warnings though) with 2.18 kernel components
    however. The 2.19 kernel patch is for 2.6.22.
  • an experimental Solaris kernel driver has been added - tested with Solaris
    11.snv only, but may work with older ones.
  • fix a file descriptor leak in rpld (thanks J.Eller)
  • fix a symlink recursion when walking /dev (J.Eller)
  • raise maximum number of open files (J.Eller)
  • handle permission restrictions on /dev nodes
  • fix directory name of Persian translation
  • moved to autotools
  • build with new libHX 1.10

v2.18 (2006-07-01)

  • Fixed OpenBSD 3.9 kpatch
  • Readded workaround for OpenBSD/NetBSD EINTR signal handling that got lost
    in 2.17

v2.17 (2006-06-19)

  • Support for Linux 2.6.17, FreeBSD 6.1, OpenBSD 3.9
  • added missing -fprofile-arcs -ftest-coverage to LDFLAGS for PROF=1

v2.16 (2006-04-01)

  • user/rpld.c: removed a superfluous variable
  • Added polling functions to the BSD rpldevs. This fixes the high CPU usage
    (which is due to the implementation of pthreads on OpenBSD) observed with
    rpld/OpenBSD.
  • OpenBSD part is again just a patch since modules are said to be unsupported
    on non-i386
  • Due to popular request, the docs now come with a black-on-white ("paper
    color") color scheme and no fixed font size. The old gray-on-black
    ("console color") is available as an alternate stylesheet and can be used
    in Firefox using View > Page Style > Black background.
  • Added the EVT_MAGIC pseudo event type
  • Added share/ttyrpld.magic for the file(1) utility

v2.15 (2006-03-11)

General:

  • added support for NetBSD 3.0
  • shifts for majors>255 and minors>255 were broken in include/dev.h

kpatch:

  • [All] fixed some SMP race conditions
  • [OpenBSD] fixed: rpl_deinit() was not called in ttyclose()

rpldev:

  • [All] rename functions to make function more clear (mv_to_user ->
    circular_get, mv_buffer -> circular_put_packet, mv_buffer2 -> circular_put)
  • [All] make avail_R() and avail_W() return size_t instead of ssize_t,
    addressing some potential overflow problems
  • [Linux] document circulary calculation of avail_*(), please report errors
  • [FreeBSD] merged freebsd-*/rpldev.c into one file
  • [FreeBSD/OpenBSD] removed SKIP_PTM, since master ptys do not generate any
    noise (that's a kpatch feature)
  • [OpenBSD] rpldev is now a module (rpldev_mod.o)
  • [OpenBSD] loads at new device number (228,0), please re-mknod your /dev/rpl
    node

v2.12 (2006-01-21)

  • fixed: FreeBSD 6.0 hands krpl_close() a NULL pointer in some cases; handle
    it properly
  • some doc updates
  • GNUmakefile: "all" target does not include "doc" anymore, because BSDs
    (standard installation) often ship without php, w3m and perl.
  • added the '[' and ']' keys to ttyreplay (lower/raise speed by 0.10)

v2.11 (2006-01-06)

Added support back in for the latest releases of the previous generation of
kernels. For Linux, that's 2.4.something (USE DISCOURAGED!), and for FreeBSD
that's 5.3.

v2.10 (2005-12-04)

Support for OpenBSD 3.8 and FreeBSD 6.0 was added; permission denied, syslog
spamming, cpu hog, ownership and symlink recursion bugs were fixed. Docs were
updated to reflect advances since the last seven releases.

  • Support for OpenBSD 3.8 and FreeBSD 6.0
  • Documentation has been updated to reflect advances since the last seven
    releases.
  • user/replay.c: fixed a color bug when -t was active (spotted whilst under
    FreeBSD)
  • user/*.c: use usleep(10000) instead of sched_yield() to reduce idle cpu hog
  • user/rpld.c: umask was set wrong, so 1.) logging did not take place 2.)
    syslog was spammed with permission denied warnings
  • user/rpld.c: added a rate limiter for syslog messages
  • user/rpld.c: find_devnode_dive() forgot to check for symlinks so that we
    dived into /dev/fd, which is not wanted, because /dev/fd/* itself includes
    /dev again, which lead to unwanted results.
  • user/rpld.c: fixed a filedescriptor leak in fill_info()
  • user/rpld.c: another try at getting ownerships right; rpld still had a case
    where the recorded owner differed from the owner of the /dev node (with
    respect to the design decision). See one below:
  • linux-2.6/rpldev.c: /dev/tty is a special case with respect to ownerships,
    so it must be filtered out in krpl_init() and krpl_open().

doc/rpl.5.php
<?php include_once("_header.php"); ?>

<h1>ttyrpld log file format</h1>

<p class="block">The logfiles <i>rpld</i> creates are a small variation of the
packet protocol as described in <a href="rpldev.4.php">rpldev(4)</a>. The
<code>-&gt;dev</code> field is not present. (Since ttyrpld v2.00, the timestamp
is already added in the kernel.) All fields are little endian and packed, i.e.
there are no alignment gaps. The structure is similar to the <code>struct
rpldev_packet</code>:</p>

<p class="code"><code>
<b>struct</b> rpldsk_packet {<br />
&nbsp; &nbsp; <b>uint16_t</b> size;<br />
&nbsp; &nbsp; <b>uint8_t</b> event, magic;<br />
&nbsp; &nbsp; <b>struct</b> timeval time;<br />
} <i>__attribute__((packed))</i>;</code></p>

<p class="block">The <code>-&gt;magic</code> is always <code>0xEE</code>.</p>
Possible values for <code>-&gt;event</code> equal those listed in <a
href="rpldev.4.php">rpldev(4)</a>, plus the following:</p>

<p class="code"><code>
<b>enum</b> {<br />
&nbsp; &nbsp; EVT_MAGIC <b>=</b> 0x4A,<br />
&nbsp; &nbsp; EVT_ID_PROG <b>=</b> 0xF0,<br />
&nbsp; &nbsp; EVT_ID_DEVPATH <b>=</b> 0xF1,<br />
&nbsp; &nbsp; EVT_ID_TIME <b>=</b> 0xF2,<br />
&nbsp; &nbsp; EVT_ID_USER <b>=</b> 0xF3,<br />
};</code></p>

<p class="block">A <code>EVT_ID_PROG</code> is added to each logfile by
<i>rpld</i> to contain the program with which it was created. (<i>rpld</i> in
this case&nbsp;-- but anyone is free to write a different daemon.)</p>

<p class="block">A <code>EVT_ID_DEVPATH</code> contains the path of the device
node that has been tapped. There is a special handling case in <i>rpld</i> when
multiple device nodes with the same major-minor number are used at the same
time:</p>

<ul>
  <li>if the first packet <i>rpld</i> picks up for a certain
    <code>-&gt;dev</code> line is a <code>EVT_INIT</code> or
    <code>EVT_OPEN</code> packet with a dentry name (e.g. 
    <code>/dev/tty1</code>), this name will be used for this major-minor
    number</li>
  <li><i>rpld</i> will look into <code>/dev</code> and take the first pick</li>
</ul>

<p class="block">In the default case, you should not worry, as each major-minor
number only has one node in <code>/dev</code>, and even if it does not, you are
able to know what tty was logged, because the dentries are similarly-named.</p>

<p class="block"><code>EVT_ID_TIME</code> packets carry the time the log was
started. This is useful because the logfile's timestamp on the filesystem may
change due to user interaction.</p>

<p class="block">Last, but not least, <code>EVT_ID_USER</code> contains the
username (or UID if no user was found) that was being traced. Especially useful
when users get removed after the logs have been recorded.</p>

<p class="block">All of these four packets are informational ones and are not
required for proper replay with <i>ttyreplay</i>.</p>

<h1>See also</h1>

<ul>
  <li><a href="rpldev.4.php"><code>rpldev(4)</code></a></li>
</ul>

<?php include_once("_footer.php"); ?>
